File: src/causal/fairness.rs
//! Fairness Objective
//!
//! An in-processing fairness objective that adds a regularization penalty
//! to the standard log-loss gradient.  Two modes are supported:
//!
//! * **Demographic Parity** — penalizes correlation between $\hat{Y}$ and $S$.
//! * **Equalized Odds** — penalizes correlation between $\hat{Y}$ and $S$
//!   *conditionally* within each class of the true label $Y$.
use crate::metrics::evaluation::Metric;
use crate::objective_functions::objective::ObjectiveFunction;
use serde::{Deserialize, Serialize};

/// Which fairness criterion to enforce.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum FairnessType {
    /// Penalize overall disparity: $\text{Mean}(\hat{p}|S\!=\!1) - \text{Mean}(\hat{p}|S\!=\!0)$.
    DemographicParity,
    /// Penalize disparity within each label class (Equalized Odds):
    /// $\sum_{y \in \{0,1\}} \bigl(\text{Mean}(\hat{p}|S\!=\!1, Y\!=\!y) - \text{Mean}(\hat{p}|S\!=\!0, Y\!=\!y)\bigr)^2$.
    EqualizedOdds,
}

/// Fairness Objective (In-processing).
///
/// Adds a regularization term to log-loss to penalize predictions that
/// correlate with a sensitive attribute $S$.
///
/// $$L = \text{LogLoss} + \lambda \cdot \text{Penalty}(\hat{Y}, S)$$
///
/// The penalty form depends on the chosen [`FairnessType`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FairnessObjective {
    /// Group membership vector (e.g., 0 or 1 for each sample).
    pub sensitive_attr: Vec<i32>,
    /// Penalty strength for the fairness regularization term.
    pub lambda: f32,
    /// Fairness criterion to enforce.
    pub fairness_type: FairnessType,
}

impl FairnessObjective {
    /// Create a new `FairnessObjective` with Demographic Parity.
    ///
    /// * `sensitive_attr` - Binary group membership for each sample.
    /// * `lambda` - Strength of the fairness penalty.
    pub fn new(sensitive_attr: Vec<i32>, lambda: f32) -> Self {
        Self {
            sensitive_attr,
            lambda,
            fairness_type: FairnessType::DemographicParity,
        }
    }

    /// Create a new `FairnessObjective` with a specific fairness type.
    pub fn with_type(sensitive_attr: Vec<i32>, lambda: f32, fairness_type: FairnessType) -> Self {
        Self {
            sensitive_attr,
            lambda,
            fairness_type,
        }
    }
}

/// Accumulator for group-conditional statistics.
#[derive(Default)]
struct GroupStats {
    sum: f64,
    count: f64,
}

impl GroupStats {
    fn mean(&self) -> f64 {
        if self.count < 1.0 { 0.0 } else { self.sum / self.count }
    }
    fn safe_count(&self) -> f64 {
        if self.count < 1.0 { 1.0 } else { self.count }
    }
}

impl ObjectiveFunction for FairnessObjective {
    fn loss(&self, y: &[f64], yhat: &[f64], _sample_weight: Option<&[f64]>, _group: Option<&[u64]>) -> Vec<f32> {
        // Standard LogLoss (the penalty is applied via gradient modification only).
        y.iter()
            .zip(yhat.iter())
            .map(|(y_i, yhat_i)| {
                let p = 1.0 / (1.0 + (-yhat_i).exp());
                let score = -(y_i * p.max(1e-15).ln() + (1.0 - y_i) * (1.0 - p).max(1e-15).ln());
                score as f32
            })
            .collect()
    }

    fn gradient(
        &self,
        y: &[f64],
        yhat: &[f64],
        _sample_weight: Option<&[f64]>,
        _group: Option<&[u64]>,
    ) -> (Vec<f32>, Option<Vec<f32>>) {
        let n = y.len();
        let mut grad = Vec::with_capacity(n);
        let mut hess = Vec::with_capacity(n);

        // Compute predicted probabilities.
        let probs: Vec<f64> = yhat.iter().map(|yh| 1.0 / (1.0 + (-yh).exp())).collect();

        // Scale fairness gradient by N to match LogLoss scale (sum over N).
        let n_f64 = n as f64;

        match &self.fairness_type {
            FairnessType::DemographicParity => {
                // Compute mean predicted probability per sensitive group.
                let mut s1 = GroupStats::default();
                let mut s0 = GroupStats::default();
                for (i, &p) in probs.iter().enumerate() {
                    if self.sensitive_attr[i] == 1 {
                        s1.sum += p;
                        s1.count += 1.0;
                    } else {
                        s0.sum += p;
                        s0.count += 1.0;
                    }
                }
                let diff = s1.mean() - s0.mean();

                for i in 0..n {
                    let p = probs[i];
                    let dp = p * (1.0 - p); // dsigma/dscore
                    let mut g = p - y[i]; // standard LogLoss gradient

                    // d(diff^2)/dscore_i
                    let fairness_grad = if self.sensitive_attr[i] == 1 {
                        2.0 * (self.lambda as f64) * diff * (1.0 / s1.safe_count()) * dp * n_f64
                    } else {
                        2.0 * (self.lambda as f64) * diff * (-1.0 / s0.safe_count()) * dp * n_f64
                    };
                    g += fairness_grad;

                    grad.push(g as f32);
                    hess.push(dp as f32);
                }
            }
            FairnessType::EqualizedOdds => {
                // Compute conditional mean per (sensitive group, true label) cell.
                // Cells: (S=0,Y=0), (S=0,Y=1), (S=1,Y=0), (S=1,Y=1)
                let mut s0_y0 = GroupStats::default();
                let mut s0_y1 = GroupStats::default();
                let mut s1_y0 = GroupStats::default();
                let mut s1_y1 = GroupStats::default();

                for i in 0..n {
                    let p = probs[i];
                    let label = if y[i] >= 0.5 { 1 } else { 0 };
                    let group = self.sensitive_attr[i];
                    match (group, label) {
                        (1, 1) => {
                            s1_y1.sum += p;
                            s1_y1.count += 1.0;
                        }
                        (1, 0) => {
                            s1_y0.sum += p;
                            s1_y0.count += 1.0;
                        }
                        (_, 1) => {
                            s0_y1.sum += p;
                            s0_y1.count += 1.0;
                        }
                        (_, 0) => {
                            s0_y0.sum += p;
                            s0_y0.count += 1.0;
                        }
                        _ => unreachable!(),
                    }
                }

                let diff_y0 = s1_y0.mean() - s0_y0.mean();
                let diff_y1 = s1_y1.mean() - s0_y1.mean();

                for i in 0..n {
                    let p = probs[i];
                    let dp = p * (1.0 - p);
                    let mut g = p - y[i]; // standard LogLoss gradient

                    let label = if y[i] >= 0.5 { 1 } else { 0 };
                    let (diff, cnt_s1, cnt_s0) = if label == 1 {
                        (diff_y1, s1_y1.safe_count(), s0_y1.safe_count())
                    } else {
                        (diff_y0, s1_y0.safe_count(), s0_y0.safe_count())
                    };

                    let fairness_grad = if self.sensitive_attr[i] == 1 {
                        2.0 * (self.lambda as f64) * diff * (1.0 / cnt_s1) * dp * n_f64
                    } else {
                        2.0 * (self.lambda as f64) * diff * (-1.0 / cnt_s0) * dp * n_f64
                    };
                    g += fairness_grad;

                    grad.push(g as f32);
                    hess.push(dp as f32);
                }
            }
        }

        (grad, Some(hess))
    }

    fn initial_value(&self, _y: &[f64], _sample_weight: Option<&[f64]>, _group: Option<&[u64]>) -> f64 {
        0.0
    }

    fn default_metric(&self) -> Metric {
        Metric::LogLoss
    }
}

File: package-python/python/perpetual/fairness.py
"""Fairness-aware gradient boosting classifier.

Provides ``FairClassifier``, a wrapper around ``PerpetualBooster`` that adds
an in-processing fairness regularization penalty to the log-loss gradient.
Two fairness criteria are supported:

* **Demographic Parity** — penalizes correlation between predictions and a
  sensitive attribute :math:`S`.
* **Equalized Odds** — penalizes correlation between predictions and
  :math:`S` *conditionally* within each class of the true label :math:`Y`.

The objective is:

.. math::

    L = \\text{LogLoss} + \\lambda \\cdot \\text{Penalty}(\\hat{Y}, S)
"""

from typing import Optional

import numpy as np
from typing_extensions import Self

from perpetual.booster import PerpetualBooster

# ---------------------------------------------------------------------------
# Custom objective helpers  (mirror Rust FairnessObjective)
# ---------------------------------------------------------------------------

_VALID_FAIRNESS_TYPES = ("demographic_parity", "equalized_odds")


def _fair_loss(sensitive_attr):
    """Return a loss callable (standard log-loss; penalty is gradient-only)."""

    def loss(y, yhat, sample_weight, group):
        y = np.asarray(y)
        yhat = np.asarray(yhat)
        p = 1.0 / (1.0 + np.exp(-yhat))
        score = -(
            y * np.log(np.maximum(p, 1e-15))
            + (1.0 - y) * np.log(np.maximum(1.0 - p, 1e-15))
        )
        return score.tolist()

    return loss


def _fair_gradient_dp(sensitive_attr, lam):
    """Gradient for Demographic Parity fairness."""
    s = np.asarray(sensitive_attr)

    def gradient(y, yhat, sample_weight, group):
        y = np.asarray(y)
        yhat = np.asarray(yhat)
        p = 1.0 / (1.0 + np.exp(-yhat))
        dp = p * (1.0 - p)

        # Group means
        mask1 = s == 1
        mask0 = ~mask1
        n1 = float(mask1.sum()) or 1.0
        n0 = float(mask0.sum()) or 1.0
        mean1 = p[mask1].sum() / n1
        mean0 = p[mask0].sum() / n0
        diff = mean1 - mean0

        grad = p - y  # standard log-loss gradient

        # Fairness penalty gradient
        fair_grad = np.where(
            mask1,
            2.0 * lam * diff * (1.0 / n1) * dp,
            2.0 * lam * diff * (-1.0 / n0) * dp,
        )
        grad = grad + fair_grad
        hess = dp

        return grad.tolist(), hess.tolist()

    return gradient


def _fair_gradient_eo(sensitive_attr, lam):
    """Gradient for Equalized Odds fairness."""
    s = np.asarray(sensitive_attr)

    def gradient(y, yhat, sample_weight, group):
        y = np.asarray(y)
        yhat = np.asarray(yhat)
        p = 1.0 / (1.0 + np.exp(-yhat))
        dp = p * (1.0 - p)
        label = (y >= 0.5).astype(int)

        grad = p - y  # standard log-loss gradient
        hess = dp.copy()

        for lbl in (0, 1):
            lbl_mask = label == lbl
            s1_lbl = (s == 1) & lbl_mask
            s0_lbl = (s == 0) & lbl_mask
            n1 = float(s1_lbl.sum()) or 1.0
            n0 = float(s0_lbl.sum()) or 1.0
            mean1 = p[s1_lbl].sum() / n1 if s1_lbl.any() else 0.0
            mean0 = p[s0_lbl].sum() / n0 if s0_lbl.any() else 0.0
            diff = mean1 - mean0

            fair_grad = np.where(
                s1_lbl,
                2.0 * lam * diff * (1.0 / n1) * dp,
                np.where(
                    s0_lbl,
                    2.0 * lam * diff * (-1.0 / n0) * dp,
                    0.0,
                ),
            )
            grad = grad + fair_grad

        return grad.tolist(), hess.tolist()

    return gradient


def _fair_initial_value():
    """Return an initial-value callable."""

    def initial_value(y, sample_weight, group):
        return 0.0

    return initial_value


# ---------------------------------------------------------------------------
# FairClassifier
# ---------------------------------------------------------------------------


class FairClassifier:
    """Fairness-aware gradient boosting classifier.

    Wraps a ``PerpetualBooster`` with an in-processing fairness penalty that
    regularizes the log-loss gradient to reduce dependence of predictions on
    a sensitive attribute.

    Parameters
    ----------
    sensitive_feature : int
        Column index of the sensitive attribute in *X*.  The column must be
        binary (0 or 1).
    fairness_type : str, default="demographic_parity"
        Fairness criterion.  One of:

        - ``"demographic_parity"`` — penalize overall disparity.
        - ``"equalized_odds"`` — penalize disparity within each label class.
    lam : float, default=1.0
        Strength of the fairness penalty (:math:`\\lambda`).
    budget : float, default=0.5
        Fitting budget forwarded to ``PerpetualBooster``.
    **kwargs
        Additional keyword arguments forwarded to ``PerpetualBooster``.

    Attributes
    ----------
    feature_importances_ : ndarray of shape (n_features,)
        Feature importances from the fitted model.

    Examples
    --------
    >>> from perpetual.fairness import FairClassifier
    >>> import numpy as np
    >>> X = np.column_stack([np.random.randn(200, 3),
    ...                      np.random.binomial(1, 0.5, 200)])
    >>> y = (X[:, 0] > 0).astype(float)
    >>> clf = FairClassifier(sensitive_feature=3, lam=2.0)
    >>> clf.fit(X, y)  # doctest: +SKIP
    >>> probs = clf.predict_proba(X)  # doctest: +SKIP

    Notes
    -----
    The fairness penalty is applied only through the gradient; the reported
    loss is standard log-loss.  This mirrors the Rust ``FairnessObjective``
    implementation.
    """

    def __init__(
        self,
        sensitive_feature: int,
        fairness_type: str = "demographic_parity",
        lam: float = 1.0,
        budget: float = 0.5,
        **kwargs,
    ):
        if fairness_type not in _VALID_FAIRNESS_TYPES:
            raise ValueError(
                f"fairness_type must be one of {_VALID_FAIRNESS_TYPES}, "
                f"got {fairness_type!r}."
            )
        self.sensitive_feature = sensitive_feature
        self.fairness_type = fairness_type
        self.lam = lam
        self.budget = budget
        self._kwargs = kwargs
        self.feature_importances_: Optional[np.ndarray] = None

    def fit(self, X, y) -> Self:
        """Fit the fair classifier.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Feature matrix.  The column at index ``sensitive_feature`` must
            contain binary (0/1) values.
        y : array-like of shape (n_samples,)
            Binary target variable (0 or 1).

        Returns
        -------
        self
            Fitted estimator.
        """
        X_arr = np.asarray(X, dtype=float)
        y_arr = np.asarray(y, dtype=float)
        sensitive_attr = X_arr[:, self.sensitive_feature].astype(int)

        if self.fairness_type == "demographic_parity":
            grad_fn = _fair_gradient_dp(sensitive_attr, self.lam)
        else:
            grad_fn = _fair_gradient_eo(sensitive_attr, self.lam)

        objective = (
            _fair_loss(sensitive_attr),
            grad_fn,
            _fair_initial_value(),
        )

        self._model = PerpetualBooster(
            budget=self.budget, objective=objective, **self._kwargs
        )
        self._model.fit(X_arr, y_arr)
        self.feature_importances_ = self._model.feature_importances_
        return self

    def predict(self, X) -> np.ndarray:
        """Predict class labels.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Feature matrix.

        Returns
        -------
        ndarray of shape (n_samples,)
            Predicted class labels (0 or 1).
        """
        scores = self._model.predict(X)
        return (scores > 0).astype(int)

    def predict_proba(self, X) -> np.ndarray:
        """Predict class probabilities.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Feature matrix.

        Returns
        -------
        ndarray of shape (n_samples, 2)
            Predicted probabilities for class 0 and class 1.
        """
        scores = self._model.predict(X)
        p1 = 1.0 / (1.0 + np.exp(-scores))
        return np.column_stack([1.0 - p1, p1])

    def decision_function(self, X) -> np.ndarray:
        """Return raw log-odds scores.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Feature matrix.

        Returns
        -------
        ndarray of shape (n_samples,)
            Raw boosted scores (log-odds).
        """
        return self._model.predict(X)
